enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,390 @@
+track of all the secrets referenced by the routes. This component is solely
+responsible for maintaining all the single item list-watch functions required
+to cache the referenced secrets.
+
+The router will bootstrap the secret monitor to ensure it can keep the
+certificate/secret up-to-date. This means the router pod will maintain active watches
+for every secret that is referenced by a route.
+
+Every active watch will be linked to a route, meaning the secret monitor
+will be linked to the lifecycle of the route. For every new route that is created,
+the secret monitor will start a watch if the route uses `.spec.tls.externalCertificate`.
+If a route is deleted, the secret monitor will deregister the route and teardown
+the watch associated with it.
+
+The `ServiceAliasConfig` creation logic will be updated in the router to also parse
+the secret referenced in `.spec.tls.externalCertificate`. The router will
+use the default certificates only when `.spec.tls.certificate/key` or `.spec.tls.externalCertificate`
+are not provided.
+
+The cluster-ingress-operator will propagate the relevant Tech-Preview feature gate down to the
+router. This feature gate will be added as a command-line argument called `ROUTER_EXTERNAL_CERTIFICATE`
+to the router and will not be user configurable.
+
+### Risks and Mitigations
+
+There is a possibility of an invalid route being processed by the router (edge case),
+if any changes are done to the rbac or the referenced secret is deleted after the API
+server validation but before router has processed the route (maybe router pod is not running)
+then this can lead to the router processing this incorrect route.
+
+> Will need to duplicate the validations present on the API server to the router.
+
+### Drawbacks
+
+The user will need to manually create, provide and maintain the rbac required by the
+router so that it can access secrets securely. This becomes tedious when users have
+many of Routes.
+
+The workaround for this is to document various levels of rbac that can be provided,
+
+- Grant router service account access to secret by secret-name (explicit rbac)
+- Grant router service account access to all secrets in a fixed namespace (implicit rbac)
+
+The above variations need to be documented for the end user as part of OpenShift documentation.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Performance testing of openshift-router in tech-preview? Is there
+  a workflow present where we can gather some early metrics (memory, cpu)? This will help in
+  preemptively addressing performance concerns before going GA.
+
+  > Will be addressed when taking feature from TP -> GA.
+
+- Do we make changes to the ingress-to-route controller as well?
+
+  > The ingress-to-route behaviour will remain as is i.e. it will not make use of
+  > the newly introduced field.
+
+- What should be the behaviour of `ValidationHostUpdate()` when using `externalCertificate`?
```

Comment:
Would be good to reference `ValidateHostUpdate` function like you did in the implementation details.

Reply:
Added a link.